### ISSUE TITLE
[strict-boolean-expressions] False positive when using union types

### ISSUE THREAD

**Issue by eamodio** [typescript-eslint/typescript-eslint#2894]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": [
      "warn",
      {
	"allowString": true,
	"allowNumber": true,
	"allowNullableObject": true,
	"allowNullableBoolean": true,
	"allowNullableNumber": false,
	"allowNullableString": true,
	"allowAny": false
      }
    ],
  }
}
```

```TS
function foo(value?: boolean | { prop: any }) {
	if (value) {
		// value above errors with "Unexpected value in conditional. A boolean expression is required."
	}
}
```

`tsconfig.json`
```json
{
	"compilerOptions": {
		"esModuleInterop": true,
		"experimentalDecorators": true,
		"forceConsistentCasingInFileNames": true,
		"incremental": true,
		"isolatedModules": true,
		"lib": ["es2019"],
		"module": "esnext",
		"moduleResolution": "node",
		"noFallthroughCasesInSwitch": true,
		"noImplicitReturns": true,
		"noUnusedLocals": false,
		"outDir": "dist",
		"resolveJsonModule": true,
		"rootDir": "src",
		"skipLibCheck": true,
		"sourceMap": true,
		"strict": true,
		"target": "es2019"
	}
}
```

**Expected Result**

No error on the `if (value) {` line

**Actual Result**

"Unexpected value in conditional. A boolean expression is required." on the `if (value) {` line

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.10.0` |
| `@typescript-eslint/parser`        | `4.10.0` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.16.0` |
| `node`                             | `12.16.2` |

**Comment by bradzacher** [comment]
Right now this is "working as intended"
The rule is built around the idea of reducing the number of truthy/falsy states that are evaluated in a condition.

`if (boolean | object | undefined) {}` has:
- 2 truthy states:
  - `true`
  - `object`
- 2 falsey states:
  - `false`
  - `undefined`

With your options - you've specified that you don't care about the `undefined` case (eg for the purposes of the rule, there's only 1 falsey case).
However this means we still have 2 truthy states.

Which begs the question: "is `true` _intentionally_ treated the same as object exists?"
They're distinct values - do they mean the same thing logically?

The question is - how should we handle this?
What's the correct behaviour? From a strictness / case handling POV - the rule is correct in flagging this.
From a user's perspective - I can definitely understand how this is a surprising result.

**Comment by eamodio** [comment]
Yeah, I definitely come down on the user's side on this one 😉 Mainly because given the settings, if the value were either of the two truthy states it would be allowed, so I would want the combination to be allowed. But if extra strictness is desire/required, maybe have a `strictUnions` option or something that would opt into that behavior?

**Comment by phaux** [comment]
It works fine in this case. This rule is all about being explicit. You should write your condition as something like `value != undefined && value !== false` and it should work. The whole reason for this rule is to make the conditions explicit about what values they check.

**Comment by aaronadamsCA** [comment]
TLDR, `strict-boolean-expressions` and `React.ReactNode` do not play together nicely.

```tsx
import type { ReactNode } from "react";

interface ComponentWithOptionalChildrenProps {
  children?: ReactNode;
}

export const ComponentWithOptionalChildren = ({
  children,
}: ComponentWithOptionalChildrenProps): JSX.Element => (
  <div>
    <h1>Component with optional children</h1>
    {children && <div>{children}</div>}
  </div>
);
```

That `children` check triggers this rule. Given a `ReactNode` can be pretty much anything (`boolean`, `string`, `number`, `null`, `undefined`, or a variety of `object`s), I do understand this is a poor fit for strict boolean checks; on the other hand I don't have much control over React node types.

Unfortunately `{Boolean(children) && <div>{children}</div>}` [breaks TypeScript narrowing](https://github.com/microsoft/TypeScript/issues/16655) with no apparent plan to fix, so it's either `{children !== undefined && children !== null && children && <div>{children}</div>}` every time, or remove this rule. For now I'm opting for the latter to keep things simple.

I don't have a suggested solution here, but I did want to mention the common use case.

**Comment by bradzacher** [comment]
Try using `!= null`. It's shorter, understood by TS and clear to developers.

`children != null && <div>{children}</div>`

At Facebook the entire codebase exclusively uses `== null` checks because in flow there is the "maybe" operator (`?boolean` means `boolean | null | undefined`). Because of the operator we *never* annotate `| null` manually, so we always have to check for undefined in code. IMO it's a huge improvement in coding style.

**Comment by aaronadamsCA** [comment]
Thanks, that's interesting; so it sounds like you don't enforce triple-equals in TS. That makes sense; I hadn't even considered that `eqeqeq` ceases to be a useful rule once you've got all these TS rules up and running in their place. I'm definitely going to give that a try.

**Comment by bradzacher** [comment]
> so it sounds like you don't enforce triple-equals in TS.

On the contrary - `eqeqeq` is enabled, but we have the option turned on which allows you to use double equals specifically for comparisons with `null`

See the end of this section: https://eslint.org/docs/rules/eqeqeq#always
```
"eqeqeq": ["error", "always", {"null": "ignore"}],
```

**Comment by aaronadamsCA** [comment]
Wow, that is 1000x better than what I was doing, thank you!

**Comment by Bhavin789** [comment]
Have confusion around the same thing. I am using union type:

```
export interface Data {
 a :string;
 b: string;
}
```

and 

```
export interface DataWithC extends Data {
 c: number;
}
```

I have a type defined like 
```
interface Props {
 property: Data[] | DataWithC[]
}
```


Now, I am using it like: 

```
const props: Props = {
 property: {
   a: "A",
   b: "B"
 }
}

const data = props.map(({a, b}) => {
 console.log(b): // Gives `b` as any type but hovering over in VSCode gives the correct type
})
```

There might be some foolish thing I am missing but any clue why might this be happening? Stuck for long on this rule. 

My expectation is that if the TS control flow analysis that `b` exists, then this rule shuld also do that. 
Thanks!

**Comment by phaux** [comment]
I believe this issue can be closed

**Comment by JoshuaKGoldberg** [comment]
Indeed. I think the rule is working as intended, as previously mentioned. There are two falsy cases in type `boolean | { prop: any } | undefined`; the intent of the rule is to stop checks on them from happening.

One could make an argument that the combination of `allowNullableBoolean` and `allowNullableObject` should stop that case, but each argument on its own is only meant to target a type union of just that type + nullable.

Marking this as working-as-intended because of all that, and that there's been very little activity on this issue in the almost 2 years it's been open. Feel free to post back or make a new issue if you think this is incorrect! Thanks all!
